[1/n][Optimus][Auto-AC] Support activation quantization without scaling

mengluy0125 opened the pull request:
Summary: We enable the activation quantization in the forward pass, and users can customize the dtype they want to quantize.

Test Plan:
# unit test

```
buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/inductor:quantization -- test_activation_quantization_aten
```

Buck UI: https://www.internalfb.com/buck2/776d3911-bb86-4ac8-a527-540cf1510b9d
Test UI: https://www.internalfb.com/intern/testinfra/testrun/4785074873051017
Network: Up: 4.3MiB  Down: 42MiB  (reSessionID-fef7e727-68b1-4645-a519-5652854df38d)
Executing actions. Remaining     0/4                                                                                 6.7s exec time total
Command: test.     Finished 2 local
Time elapsed: 3:11.5s
Tests finished: Pass 2. Fail 0. Fatal 0. Skip 0. Build failure 0

# E2E

### how to enable (you can overrite the dtype, if nothing given, the default is fp8)

```
post_grad_fusion_options={
            "activation_quantization_aten_pass": {"quant_type": "torch.float8_e5m2"}
        },
```

Differential Revision: D70522237




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

eellison commented on torch/_dynamo/utils.py:
```diff
@@ -4588,3 +4588,7 @@ def maybe_disable_inference_mode_for_fake_prop() -> Generator[None, None, None]:
         yield
+
+
+def is_node_meta_valid(node: Optional[torch.fx.Node]):
```
nit: -> `bool`

mengluy0125 replied:
I have to move part of the functions in torch/_inductor/pattern_matcher.py to torch/_functorch/pattern_matcher.py due to the circular dependency problem. For any existing dependency on torch/_inductor/pattern_matcher.py, I will redirect it to torch/_functorch/pattern_matcher.py by adding the "dummy" module imports there to avoid current dependency break.

eellison replied:
What exactly is the circular dependency ?

eellison replied:
```
@register_graph_pattern(
    Placeholder("*", users=MULTIPLE),
    pass_dict=construct_pattern_matcher_pass("activation_quantization_fwd_aten_pass"),
)
```

As mentioned earlier - this should just be `graph.find_nodes`. Which will be O(1) instead of O(n) where n is the number of graphs in the node. 

If the only reason for all the churn is to do the above, which is slower, would prefer not to.

eellison replied:
you can also do something like
```
def foo():
     import ....


foo()
```

to register and avoid the refactoring

mengluy0125 replied:
Sure, I can do graph.find_nodes instead, then we don't need to do the code refactoring @eellison

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -296,12 +322,109 @@ def _remove_by_name(saved_values: list[fx.Node], name: str):
+):
+    GraphTransformObserver = functools.partial(
+        torch.fx.passes.graph_transform_observer.GraphTransformObserver,
+        subsystem="default_partition_passes",
```
The subsystem is primarily to bisect application of passes with CompilerBisector. https://github.com/pytorch/pytorch/blob/fd3d339e17e147f1bbe9f8e51ed45a9f56b0025b/torch/_inductor/compiler_bisector.py#L61-L64.

Should we add there ?

mengluy0125 replied:
SG

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -296,12 +322,109 @@ def _remove_by_name(saved_values: list[fx.Node], name: str):
+        subsystem="default_partition_passes",
+    )
+
+    if inductor_config.pattern_matcher:
```
nit: 
```
if not inductor_config.pattern_matcher:
     return
```

to reduce nesting

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -2141,3 +2277,95 @@ def draw_graph(
+
+
+def should_quantize(node: torch.fx.Node) -> bool:
+    allowed_dtypes = [
```
could also do: `dtype.is_floating_point and dtype.itemsize > torch.float8_e5m2.itemsize`

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -2141,3 +2277,95 @@ def draw_graph(
+    return (
+        is_node_meta_valid(node)
+        and node.meta["val"].dtype in allowed_dtypes
+        and len(shape) > 1
+        and shape[0] >= 1024
```
Is there a reason we only quantize if batch size >= 1024 ? as opposed to looking at total numel or something ?

mengluy0125 replied:
It's model dependent, this is the heuristic borrowed by some experiment on IGCTR. In my followup PR, I would give the flexibility to customers to set the quantize constraints themselves instead of hardcoding the requirements, thus they can customize their needs to find the best quantization condition.

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -1893,94 +1896,6 @@ def wrap_tensors(x):
-        # Complex views are supported with IR ComplexView
-        if parent and parent.target in (
-            torch.ops.aten.view.dtype,
-            torch.ops.prims.convert_element_type.default,
-        ):
-            return False
-        _warn_complex_not_supported()
-        return True
-
-    if t.is_meta:
-        return True
-
-    if t.dtype == torch.float8_e8m0fnu:
-        if not node:
-            return True
-
-        # allow bitcast, views, memory movement, but not arithmetic
-        # TODO: delete once triton adds native support
-        return not (
-            isinstance(parent.target, torch._ops.OpOverload)
-            and parent.target
-            in (
-                aten.view.dtype,
-                aten.cat.default,
-                aten._scaled_mm.default,
-            )
-            or (isinstance(node.target, torch._ops.OpOverload) and is_view(node.target))
-        )
-
-    return False
-
-
-def unsupported_output_tensor(t: torch.Tensor, parent=None, node=None):
-    "Do not support writing tensor but can read from it"
-    if unsupported_input_tensor(t, parent):
-        return True
-    return t.is_cpu and config.disable_cpp_codegen
-
-
-def fallback_node_due_to_unsupported_type(node: torch.fx.Node, allow_cpu_inputs=True):
```
Reason why this was deleted ?

mengluy0125 replied:
It is not deleted but moving it to the torch/_inductor/pattern_matcher.py due to the circular dependency problem.

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -2036,7 +2167,11 @@ def classify_nodes(joint_module, static_lifetime_input_indices):
+            joint_module,
+            _joint_inputs,
+            num_fwd_outputs=num_fwd_outputs,
+            static_lifetime_input_indices=static_lifetime_input_indices,
+            static_lifetime_input_nodes=node_info.static_lifetime_input_nodes,
```
It's kind of strange to me that we have two different ways of representing the same thing here. Is there a reason for that ? cc @bdhirsh

bdhirsh replied:
I agree, cc @mengluy0125 . Why do we need a separate concept here?

mengluy0125 replied:
I defined my enable quantization function as follows, where I use the static_lifetime_input_nodes to check if this saved node from auto_ac should be included or excluded in the quantization @bdhirsh @eellison 

```
def enable_activation_quantization(
    saved_values: list[fx.Node],
    fwd_module: fx.GraphModule,
    bwd_module: fx.GraphModule,
    static_lifetime_input_nodes: Optional[OrderedSet[fx.Node]] = None,
):
```

eellison reviewed (commented):
Haven't done full review yet. I see that _inductor/pattern_matcher.py was moved to functorch/pattern_matcher.py. 

For one, we haven't been especially dogmatic about dependencies from one subdirectory going into the other.. So I don't know if we need the move. If we do it, we need to import `torch/_functorch/pattern_matcher.py -> torch/_inductor/pattern_matcher.py` to avoid breaking BC of anyone relying on the old file. I know users such as vllm rely on it.

If we do want to do it, could we stack this as a separate refactoring pr? since additional code was added to `pattern_matcher.py` it makes it harder to review.

still have not done full review - cc @bdhirsh as well

eellison reviewed (commented):
existing comments not addressed..

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -296,12 +322,109 @@ def _remove_by_name(saved_values: list[fx.Node], name: str):
+    bwd_module: fx.GraphModule,
+    static_lifetime_input_nodes: Optional[OrderedSet[fx.Node]] = None,
+):
+    GraphTransformObserver = functools.partial(
```
nit: this is a function, not a class - I would stick with function/variable naming, e.g. `graph_transform_observer`?

mengluy0125 replied:
I followed the naming convention in the codebase to keep it consistent lol @bdhirsh 

https://github.com/pytorch/pytorch/blob/ad9e209ea387bad2a569d9acd3566c8546252980/torch/_inductor/fx_passes/joint_graph.py#L525

mengluy0125 commented:
To improve the readability, I will make a separate PR to refactor the code to resolve the circular dependency problem @eellison @bdhirsh

mengluy0125 committed:
[1/n][Optimus][Auto-AC] Support activation quantization without scaling (#148380)

Summary:
X-link: https://github.com/pytorch/benchmark/pull/2607


We enable the activation quantization in the forward pass, and users can customize the dtype they want to quantize.

Test Plan:
# unit test

```
buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/inductor:quantization -- test_activation_quantization_aten
```

Buck UI: https://www.internalfb.com/buck2/9a53c909-d3ea-479a-874e-cc917999ca88
Test UI: https://www.internalfb.com/intern/testinfra/testrun/12384899050440719
Network: Up: 62KiB  Down: 81KiB  (reSessionID-913ca82d-c395-4492-818e-6e004df37f87)
Executing actions. Remaining     0/4                                                                                                   6.1s exec time total
Command: test.     Finished 2 local
Time elapsed: 3:22.9s
Tests finished: Pass 2. Fail 0. Fatal 0. Skip 0. Build failure 0

# E2E

### how to enable

```
post_grad_fusion_options={
            "activation_quantization_aten_pass": {"quant_type": "torch.float8_e5m2"},
        },
```
see D51860030 to check how to set the config under dynamo_config_map

Note: you can change the quant_type, if nothing gives, then the default type torch.float8_e5m2 will be used to quantize

#### If you use FSDP

- You may also need to set inline_inbuilt_nn_modules to true for models that use FSDP (see D70023488 to check the config setting)
- Remove UNSAFE_SKIP_FSDP_MODULE_GUARDS=1
(context: https://fb.workplace.com/groups/1075192433118967/permalink/1629608671010671/)

```
buck2 run mode/opt //aps_models/ads/icvr:icvr_launcher -- mode=mast_omnifm_v1-5_mwb launcher.max_retries=3 data_loader.dataset.batch_size=8 launcher.data_project=oncall_ads_model_platform launcher.fbl_entitlement=ads_global_tc_training_efficiency_qps max_ind_range=1 launcher.num_workers=8 data_loader.reading_service.num_remote_dpp_workers=30 data_loader.dataset.num_batches=100 trainer.gpu_tracer.wait=50 trainer.gpu_tracer.active=3 trainer.gpu_tracer.overhead_detection=10 launcher.tags=[ads_ranking_taxonomy_mc_qps_optimization]
```

aps-512_8_remove_fsdp_guards-92ae3972ba

tlparse: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-512_8_remove_fsdp_guards-92ae3972ba/attempt_0/version_0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000


baseline w/o fp8 quantization: aps-mengluy_remove_fsdp-ce75b306fa

w/ fp8 quantization: aps-mengluy_remove_fsdp_fp8-96541deec4

### QPS

 {F1977040587}

### memory

baseline
{F1977040640}

memory snapshot:
https://www.internalfb.com/ai_infra/zoomer/profiling-run/insights?profilingRunID=1767027467197075&tab=INSIGHTS&primarySubtab=Memory%20Analysis&secondarySubtab=Memory%20Snapshot

with fp8
{F1977040641}

memory snapshot:
https://www.internalfb.com/ai_infra/zoomer/profiling-run/insights?profilingRunID=639378375763157&tab=INSIGHTS&primarySubtab=Memory%20Analysis&secondarySubtab=Memory%20Snapshot


### conclusion:

- ~9% qps improvement, reduces peak memory from 82.01 to 78.97.

- for NE, we need have longer verification, WIP with scaling version.

Reviewed By: avicizhu

Differential Revision: D70522237
